File: app/python/MusicPlayer/view_models/personalized_newsong_view_model.py
import asyncio
from enum import IntEnum
from typing import List, Dict, Any

from PyQt5.QtCore import QAbstractListModel, QModelIndex, pyqtProperty
from PyQt5.QtQuick import QQuickItem

from MusicPlayer.api.api import API
from MusicPlayer.api.data.personalized_newsong import APIPersonalizedNewSongResultData, APIPersonalizedNewSongData
from MusicPlayer.object_graph import get_object_graph
from MusicPlayer.view_models.base import BaseViewModel


class _Dependency:
    def __init__(self, api: API):
        self.api = api


class _Roles(IntEnum):
    imageUrl = 0
    id = 1
    name = 2
    nameAlias = 3
    isSq = 4
    mvId = 5
    artists = 6
    isCurrentSong = 7
    isCurrentSongPlaying = 8


class PersonalizedNewSongListModel(QAbstractListModel):
    _list: List[APIPersonalizedNewSongResultData]

    def __init__(self, parent=None):
        super().__init__(parent)
        self._list = []

    def roleNames(self) -> Dict[int, 'QByteArray']:
        roles = dict()

        for name, member in _Roles.__members__.items():
            roles[member.value] = name.encode()

        return roles

    def rowCount(self, parent: QModelIndex) -> int:
        return len(self._list)

    def reset_api_data(self, data: APIPersonalizedNewSongData):
        self.beginResetModel()
        self._list.clear()
        self._list.extend(data.result)
        self.endResetModel()

    def data(self, index: QModelIndex, role: int) -> Any:
        item = self._list[index.row()]

        if role == _Roles.imageUrl:
            return item.song.album.pic_url
        elif role == _Roles.name:
            return item.song.name
        elif role == _Roles.id:
            return item.song.id
        elif role == _Roles.nameAlias:
            return ','.join(item.song.alias)
        elif role == _Roles.artists:
            return ' / '.join([a.name for a in item.song.artists])
        elif role == _Roles.isCurrentSong:
            return False
        elif role == _Roles.isCurrentSongPlaying:
            return False
        elif role == _Roles.isSq:
            return item.song.privilege.maxbr == 999000  # 最大质量格式
        elif role == _Roles.mvId:
            return item.song.mvid


class PersonalizedNewSongViewModel(BaseViewModel):
    _d: _Dependency
    _list_model: PersonalizedNewSongListModel

    def __init__(self, parent=None):
        super().__init__(_Dependency, parent)
        self._list_model = PersonalizedNewSongListModel(self)

    def componentComplete(self):
        asyncio.create_task(self.load_data())
        super().componentComplete()

    async def load_data(self):
        data = await self._d.api.personalized_newsong()
        self._list_model.reset_api_data(data)

    @pyqtProperty(QAbstractListModel)
    def newSongListModel(self):
        return self._list_model

File: app/cpp/MusicPlayer/view_models/personalized_new_song_view_model.cc
//
// Created by rick on 2019/11/10.
//

#include "personalized_new_song_view_model.h"
#include "../util/executor.h"

namespace MusicPlayer::ViewModels {
    using namespace Util;
    PersonalizedNewSongListModel::PersonalizedNewSongListModel(QObject* parent) : QAbstractListModel{parent} {}

    QHash<int, QByteArray> PersonalizedNewSongListModel::roleNames() const {
        QHash<int, QByteArray> names;
        names[Roles::ImageUrl]             = "imageUrl";
        names[Roles::Artists]              = "artists";
        names[Roles::Id]                   = "id";
        names[Roles::IsCurrentSong]        = "isCurrentSong";
        names[Roles::IsSq]                 = "isSq";
        names[Roles::IsCurrentSongPlaying] = "isCurrentSongPlaying";
        names[Roles::MvId]                 = "mvId";
        names[Roles::Name]                 = "name";
        names[Roles::NameAlias]            = "nameAlias";
        return names;
    }

    int PersonalizedNewSongListModel::rowCount(const QModelIndex& index) const { return _data.size(); }

    QVariant PersonalizedNewSongListModel::data(const QModelIndex& index, int role) const {
        const auto& item = _data[index.row()];
        if (role == Roles::ImageUrl) {
            return item.song.album.picUrl;
        }
        if (role == Roles::Name)
            return item.song.name;
        if (role == Roles::Id)
            return item.song.id;
        if (role == Roles::NameAlias)
            return QStringList(item.song.alias.toList()).join(',');
        if (role == Roles::Artists) {
            QStringList artistNames;
            std::transform(
                item.song.artists.cbegin(), item.song.artists.cend(), std::back_inserter(artistNames), [](const auto& artist) {
                    return artist.template value<APIPersonalizedNewSongResultSongArtistData>().name;
                });
            return artistNames.join(" / ");
        }
        if (role == Roles::IsCurrentSong)
            return false;
        if (role == Roles::IsCurrentSongPlaying)
            return false;
        if (role == Roles::IsSq)
            return item.song.privilege.maxbr == 999000; // 最大质量格式
        if (role == Roles::MvId)
            return item.song.mvid;

        return {};
    }

    void PersonalizedNewSongListModel::setPersonalizedNewSongData(const APIPersonalizedNewSongData& data) {
        beginResetModel();
        _data.clear();
        std::transform(data.result.cbegin(), data.result.cend(), std::back_inserter(_data), [](const auto& var) {
            return var.template value<APIPersonalizedNewSongResultData>();
        });
        endResetModel();
    }

    PersonalizedNewSongViewModel::PersonalizedNewSongViewModel(QObject* parent) : QObject{parent} {
        _newSongModel = new PersonalizedNewSongListModel(this);
    }

    void PersonalizedNewSongViewModel::loadData() {
        if (_loading && !_loading->isReady())
            return;
        MusicAPI api;
        _loading = api.personalizedNewSong()
                       .via(AppExecutor::instance()->getMainExecutor().get())
                       .thenValue([this](const Response<APIPersonalizedNewSongData>& reply) {
                           std::visit(
                               [=](const auto& value) {
                                   if constexpr (std::is_convertible_v<decltype(value), APIPersonalizedNewSongData>) {
                                       _newSongModel->setPersonalizedNewSongData(value);
                                   } else {
                                       // errors
                                   }
                               },
                               reply);
                           return std::nullopt;
                       });
    }

    QAbstractListModel* PersonalizedNewSongViewModel::newSongListModel() const { return _newSongModel; }
} // namespace MusicPlayer::ViewModels
